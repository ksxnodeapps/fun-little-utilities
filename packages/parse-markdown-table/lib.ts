import { Stream, iterateLines, trimmedChunks } from 'string-stream-utils'
import { ArrayTable, ObjectTable, createObjectTable, createObjectTableSync } from 'table-parser-base'

export { Stream }
export { ArrayTable, ObjectTable }
export * from 'table-parser-base'

export interface TableIterationOptions {
  readonly leftBound: boolean
  readonly rightBound: boolean
}

const horizontalLineRegex = /^[|-]+$/

export async function* iterateRows(lines: Stream, options: TableIterationOptions) {
  const { leftBound, rightBound } = options

  for await (const line of lines) {
    const trimmedLine = line.trim()
    if (horizontalLineRegex.test(trimmedLine)) continue

    const row = trimmedLine.split('|').map(x => x.trim())

    if (leftBound && row.shift() !== '') {
      throw new SyntaxError('Inconsistent left bound')
    }

    if (rightBound && row.pop() !== '') {
      throw new SyntaxError('Inconsistent right bound')
    }

    yield row
  }
}

export function* iterateRowsSync(lines: Iterable<string>, options: TableIterationOptions) {
  const { leftBound, rightBound } = options

  for (const line of lines) {
    const trimmedLine = line.trim()
    if (horizontalLineRegex.test(trimmedLine)) continue

    const row = trimmedLine.split('|').map(x => x.trim())

    if (leftBound && row.shift() !== '') {
      throw new SyntaxError('Inconsistent left bound')
    }

    if (rightBound && row.pop() !== '') {
      throw new SyntaxError('Inconsistent right bound')
    }

    yield row
  }
}

export class MarkdownCellTable implements ArrayTable<string, string> {
  constructor(
    public readonly headers: readonly string[],
    public readonly rows: AsyncIterableIterator<readonly string[]>,
  ) {}
}

export class MarkdownCellTableSync implements ArrayTable<string, string> {
  constructor(
    public readonly headers: readonly string[],
    public readonly rows: IterableIterator<readonly string[]>,
  ) {}
}

export async function createMarkdownArrayTable(stream: Stream) {
  const lineIterator = trimmedChunks(iterateLines(stream))
  const firstLineResult = await lineIterator.next()

  if (firstLineResult.done) {
    throw new SyntaxError('Text is empty')
  }

  const headers = firstLineResult.value.split('|').map(x => x.trim())
  const leftBound = headers[0] === ''
  const rightBound = headers[headers.length - 1] === ''
  if (leftBound) headers.shift()
  if (rightBound) headers.pop()

  const rows = iterateRows(lineIterator, { leftBound, rightBound })

  return new MarkdownCellTable(headers, rows)
}

export function createMarkdownArrayTableSync(text: string) {
  const lines = [...text.split('\n')].map(s => s.trim())
  if (lines[0] === '') lines.shift()
  if (lines[lines.length - 1] === '') lines.pop()

  if (lines.length === 0) {
    throw new SyntaxError('Text is empty')
  }

  const headers = lines[0].split('|').map(x => x.trim())
  const leftBound = headers[0] === ''
  const rightBound = headers[headers.length - 1] === ''
  if (leftBound) headers.shift()
  if (rightBound) headers.pop()

  const rows = iterateRowsSync(lines.slice(1), { leftBound, rightBound })

  return new MarkdownCellTableSync(headers, rows)
}

export async function createMarkdownObjectTable(stream: Stream) {
  return createObjectTable(await createMarkdownArrayTable(stream))
}

export function createMarkdownObjectTableSync(text: string) {
  return createObjectTableSync(createMarkdownArrayTableSync(text))
}
