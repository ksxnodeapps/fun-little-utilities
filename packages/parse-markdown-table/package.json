{
  "name": "parse-markdown-table",
  "version": "0.0.6",
  "description": "Convert a markdown table from text to JavaScript objects",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "keywords": [
    "table",
    "markdown"
  ],
  "sideEffects": false,
  "browser": "index.mjs",
  "module": "index.mjs",
  "main": "index.js",
  "dependencies": {
    "tslib": "^2.0.1",
    "string-stream-utils": "^0.0.4",
    "table-parser-base": "^0.0.5",
    "@types/node": "^16.0.0"
  }
}
