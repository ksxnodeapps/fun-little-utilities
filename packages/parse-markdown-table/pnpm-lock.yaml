dependencies:
  '@types/node': 13.11.1
  string-stream-utils: 'link:../string-stream-utils'
  table-parser-base: 'link:../table-parser-base'
  tslib: 2.0.0
lockfileVersion: 5.1
packages:
  /@types/node/13.11.1:
    dev: false
    resolution:
      integrity: sha512-eWQGP3qtxwL8FGneRrC5DwrJLGN4/dH1clNTuLfN81HCrxVtxRjygDTUoZJ5ASlDEeo0ppYFQjQIlXhtXpOn6g==
  /tslib/2.0.0:
    dev: false
    resolution:
      integrity: sha512-lTqkx847PI7xEDYJntxZH89L2/aXInsyF2luSafe/+0fHOMjlBNXdH6th7f70qxLDhul7KZK0zC8V5ZIyHl0/g==
specifiers:
  '@types/node': ^13.11.1
  string-stream-utils: ^0.0.3
  table-parser-base: ^0.0.3
  tslib: ^2.0.0
