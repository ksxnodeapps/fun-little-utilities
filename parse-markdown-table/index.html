<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>parse-markdown-table — Reference</title>
	<meta name="description" content="Documentation for parse-markdown-table — Reference">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="assets/css/main.css">
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="assets/js/search.json" data-base=".">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="index.html" class="title">parse-markdown-table — Reference</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="globals.html">Globals</a>
				</li>
			</ul>
			<h1>parse-markdown-table — Reference</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<div class="tsd-panel tsd-typography">
				<a href="#parse-markdown-table" id="parse-markdown-table" style="color: inherit; text-decoration: none;">
					<h1>Parse Markdown Table</h1>
				</a>
				<p>Convert a markdown table from text to JavaScript objects</p>
				<a href="#usage" id="usage" style="color: inherit; text-decoration: none;">
					<h2>Usage</h2>
				</a>
				<a href="#arraytable" id="arraytable" style="color: inherit; text-decoration: none;">
					<h3><code>ArrayTable</code></h3>
				</a>
				<pre><code class="language-javascript"><span class="hljs-keyword">const</span> { createMarkdownArrayTable } = <span class="hljs-keyword">await</span> <span class="hljs-keyword">import</span>(<span class="hljs-string">'parse-markdown-table'</span>)

<span class="hljs-keyword">const</span> markdown = <span class="hljs-string">`
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
`</span>

<span class="hljs-comment">// this function can take string, string[], Iterable&lt;string&gt;, and AsyncIterable&lt;string&gt;</span>
<span class="hljs-keyword">const</span> table = <span class="hljs-keyword">await</span> createMarkdownArrayTable(markdown)

<span class="hljs-built_in">console</span>.info(<span class="hljs-string">'headers'</span>, table.headers)
<span class="hljs-keyword">for</span> <span class="hljs-keyword">await</span> (<span class="hljs-keyword">const</span> row <span class="hljs-keyword">of</span> table.rows) {
  <span class="hljs-built_in">console</span>.info(<span class="hljs-string">'row'</span>, row)
}</code></pre>
				<p><strong>Output:</strong></p>
				<pre><code class="language-javascript">headers [ <span class="hljs-string">'id'</span>, <span class="hljs-string">'name'</span>, <span class="hljs-string">'email'</span> ]
row [ <span class="hljs-string">'1'</span>, <span class="hljs-string">'John Doe'</span>, <span class="hljs-string">'john-doe@gmail.com'</span> ]
row [ <span class="hljs-string">'2'</span>, <span class="hljs-string">'Peter Smith'</span>, <span class="hljs-string">'petersmith22@outlook.com'</span> ]
row [ <span class="hljs-string">'3'</span>, <span class="hljs-string">'Julia Jones'</span>, <span class="hljs-string">'jjones778@gmail.com'</span> ]</code></pre>
				<a href="#objecttable" id="objecttable" style="color: inherit; text-decoration: none;">
					<h3><code>ObjectTable</code></h3>
				</a>
				<pre><code class="language-javascript"><span class="hljs-keyword">const</span> { createMarkdownObjectTable } = <span class="hljs-keyword">await</span> <span class="hljs-keyword">import</span>(<span class="hljs-string">'parse-markdown-table'</span>)

<span class="hljs-keyword">const</span> markdown = <span class="hljs-string">`
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
`</span>

<span class="hljs-comment">// this function can take string, string[], Iterable&lt;string&gt;, or AsyncIterable&lt;string&gt;</span>
<span class="hljs-keyword">const</span> table = createMarkdownObjectTable(markdown)

<span class="hljs-keyword">for</span> <span class="hljs-keyword">await</span> (<span class="hljs-keyword">const</span> row <span class="hljs-keyword">of</span> table) {
  <span class="hljs-built_in">console</span>.info(<span class="hljs-string">'row'</span>, row)
}</code></pre>
				<p><strong>Output:</strong></p>
				<pre><code class="language-javascript">row { <span class="hljs-attr">id</span>: <span class="hljs-string">'1'</span>, <span class="hljs-attr">name</span>: <span class="hljs-string">'John Doe'</span>, <span class="hljs-attr">email</span>: <span class="hljs-string">'john-doe@gmail.com'</span> }
row { <span class="hljs-attr">id</span>: <span class="hljs-string">'2'</span>, <span class="hljs-attr">name</span>: <span class="hljs-string">'Peter Smith'</span>, <span class="hljs-attr">email</span>: <span class="hljs-string">'petersmith22@outlook.com'</span> }
row { <span class="hljs-attr">id</span>: <span class="hljs-string">'3'</span>, <span class="hljs-attr">name</span>: <span class="hljs-string">'Julia Jones'</span>, <span class="hljs-attr">email</span>: <span class="hljs-string">'jjones778@gmail.com'</span> }</code></pre>
				<a href="#license" id="license" style="color: inherit; text-decoration: none;">
					<h2>License</h2>
				</a>
				<p><a href="https://git.io/fxKXN">MIT</a> © <a href="https://github.com/KSXGitHub">Hoàng Văn Khải</a></p>
			</div>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class="globals  ">
						<a href="globals.html"><em>Globals</em></a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
					<li class=" tsd-kind-class tsd-is-external">
						<a href="classes/markdowncelltable.html" class="tsd-kind-icon">Markdown<wbr>Cell<wbr>Table</a>
					</li>
					<li class=" tsd-kind-class tsd-is-external">
						<a href="classes/markdowncelltablesync.html" class="tsd-kind-icon">Markdown<wbr>Cell<wbr>Table<wbr>Sync</a>
					</li>
					<li class=" tsd-kind-interface tsd-is-external">
						<a href="interfaces/tableiterationoptions.html" class="tsd-kind-icon">Table<wbr>Iteration<wbr>Options</a>
					</li>
					<li class=" tsd-kind-variable tsd-is-external">
						<a href="globals.html#horizontallineregex" class="tsd-kind-icon">horizontal<wbr>Line<wbr>Regex</a>
					</li>
					<li class=" tsd-kind-function tsd-is-external">
						<a href="globals.html#createmarkdownarraytable" class="tsd-kind-icon">create<wbr>Markdown<wbr>Array<wbr>Table</a>
					</li>
					<li class=" tsd-kind-function tsd-is-external">
						<a href="globals.html#createmarkdownarraytablesync" class="tsd-kind-icon">create<wbr>Markdown<wbr>Array<wbr>Table<wbr>Sync</a>
					</li>
					<li class=" tsd-kind-function tsd-is-external">
						<a href="globals.html#createmarkdownobjecttable" class="tsd-kind-icon">create<wbr>Markdown<wbr>Object<wbr>Table</a>
					</li>
					<li class=" tsd-kind-function tsd-is-external">
						<a href="globals.html#createmarkdownobjecttablesync" class="tsd-kind-icon">create<wbr>Markdown<wbr>Object<wbr>Table<wbr>Sync</a>
					</li>
					<li class=" tsd-kind-function tsd-is-external">
						<a href="globals.html#iteraterows" class="tsd-kind-icon">iterate<wbr>Rows</a>
					</li>
					<li class=" tsd-kind-function tsd-is-external">
						<a href="globals.html#iteraterowssync" class="tsd-kind-icon">iterate<wbr>Rows<wbr>Sync</a>
					</li>
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="assets/js/main.js"></script>
</body>
</html>