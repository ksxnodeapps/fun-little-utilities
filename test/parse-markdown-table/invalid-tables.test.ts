import { tryExec } from '@tsfun/result'
import { createMarkdownObjectTable, createMarkdownObjectTableSync } from 'parse-markdown-table'
import { getAsyncArray } from './lib/async-array'

describe('async', () => {
  const setup = async (text: string) => getAsyncArray(await createMarkdownObjectTable(text))

  describe('empty string', () => {
    const text = ''

    it('rejects with a SyntaxError', async () => {
      await expect(setup(text)).rejects.toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', async () => {
      await expect(setup(text)).rejects.toMatchSnapshot()
    })
  })

  describe('string with naught but whitespaces', () => {
    const text = '  \n  \t '

    it('rejects with a SyntaxError', async () => {
      await expect(setup(text)).rejects.toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', async () => {
      await expect(setup(text)).rejects.toMatchSnapshot()
    })
  })

  describe('inconsistent right bound', () => {
    const text = `
    | id | name        | email                    |
    |----|-------------|--------------------------|
    |  1 | John Doe    | john-doe@gmail.com
    |  2 | Peter Smith | petersmith22@outlook.com
    |  3 | Julia Jones | jjones778@gmail.com
  `

    it('rejects with a SyntaxError', async () => {
      await expect(setup(text)).rejects.toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', async () => {
      await expect(setup(text)).rejects.toMatchSnapshot()
    })
  })

  describe('inconsistent left bound', () => {
    const text = `
    | id | name        | email                    |
    |----|-------------|--------------------------|
       1 | John Doe    | john-doe@gmail.com       |
       2 | Peter Smith | petersmith22@outlook.com |
    |  3 | Julia Jones | jjones778@gmail.com      |
  `

    it('rejects with a SyntaxError', async () => {
      await expect(setup(text)).rejects.toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', async () => {
      await expect(setup(text)).rejects.toMatchSnapshot()
    })
  })
})

describe('sync', () => {
  const setup = (text: string) => tryExec(() => Array.from(createMarkdownObjectTableSync(text)))

  describe('empty string', () => {
    const text = ''

    it('throws a SyntaxError', () => {
      expect(setup(text).error).toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', () => {
      expect(setup(text).error).toMatchSnapshot()
    })
  })

  describe('string with naught but whitespaces', () => {
    const text = '  \n  \t '

    it('throws a SyntaxError', () => {
      expect(setup(text).error).toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', () => {
      expect(setup(text).error).toMatchSnapshot()
    })
  })

  describe('inconsistent right bound', () => {
    const text = `
    | id | name        | email                    |
    |----|-------------|--------------------------|
    |  1 | John Doe    | john-doe@gmail.com
    |  2 | Peter Smith | petersmith22@outlook.com
    |  3 | Julia Jones | jjones778@gmail.com
  `

    it('throws a SyntaxError', async () => {
      expect(setup(text).error).toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', () => {
      expect(setup(text).error).toMatchSnapshot()
    })
  })

  describe('inconsistent left bound', () => {
    const text = `
    | id | name        | email                    |
    |----|-------------|--------------------------|
       1 | John Doe    | john-doe@gmail.com       |
       2 | Peter Smith | petersmith22@outlook.com |
    |  3 | Julia Jones | jjones778@gmail.com      |
  `

    it('throws a SyntaxError', () => {
      expect(setup(text).error).toBeInstanceOf(SyntaxError)
    })

    it('matches snapshot', () => {
      expect(setup(text).error).toMatchSnapshot()
    })
  })
})
