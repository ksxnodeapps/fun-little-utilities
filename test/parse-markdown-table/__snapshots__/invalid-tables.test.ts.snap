// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async empty string matches snapshot 1`] = `[SyntaxError: Text is empty]`;

exports[`async inconsistent left bound matches snapshot 1`] = `[SyntaxError: Inconsistent left bound]`;

exports[`async inconsistent right bound matches snapshot 1`] = `[SyntaxError: Inconsistent right bound]`;

exports[`async string with naught but whitespaces matches snapshot 1`] = `[SyntaxError: Text is empty]`;

exports[`sync empty string matches snapshot 1`] = `[SyntaxError: Text is empty]`;

exports[`sync inconsistent left bound matches snapshot 1`] = `[SyntaxError: Inconsistent left bound]`;

exports[`sync inconsistent right bound matches snapshot 1`] = `[SyntaxError: Inconsistent right bound]`;

exports[`sync string with naught but whitespaces matches snapshot 1`] = `[SyntaxError: Text is empty]`;
